### PR TITLE
Add screencopy measurement tool

### PR DESCRIPTION
**Pull request by wmww**
Based on #10. Adds a python-based tool that measures how many pixels are pushed over the screencopy protocol. Screencopy bandwidth tests are updated to record it's output as junit properties.

### CODE REVIEW

**Review by Saviq — CHANGES_REQUESTED** [review]

**Review by Saviq — COMMENTED** [review]
I added app name handling for an easier read of the test output :)

File: frame/test_screencopy_bandwidth.py [hunk 1]
```diff
@@ -26,29 +27,36 @@ def env(self) -> dict[str, str]:
...
     @pytest.mark.parametrize('server', all_servers())
-    def test_compositor_alone(self, server) -> None:
-        with DisplayServer(server) as s:
-            time.sleep(long_wait_time)
+    def test_compositor_alone(self, record_property, server) -> None:
+        with DisplayServer(server, add_extensions=ScreencopyTracker.required_extensions) as s:
+            with ScreencopyTracker(s.display_name) as tracker:
+                time.sleep(long_wait_time)
+            for name, val in tracker.properties().items():
+                record_property(name, val)
```
Comment (Saviq) [inline frame/test_screencopy_bandwidth.py:49]: Nah, that you shouldn't do. Sure, python doesn't have scopes, but it does have a garbage collector ;)

And the context manager is a… scope, of sorts.

Comment (wmww) [reply frame/test_screencopy_bandwidth.py:49]: `__exit__()` gets called at the end of the `width` block, but `tracker` is still accessible and won't be arguable collected yet. `tracker.properties()` is written to be safe to call after `__exit__()`. Doing so is a good idea to make sure properties are retrieved after all frames have been processed.

Comment (wmww) [reply frame/test_screencopy_bandwidth.py:49]: I'm under the impression it is specified behavior, but coming up with a source one way or the other is proving nontrivial, so I will do the more obvious thing and change it back.

**Comment by wmww** [comment]
I refactored out `wayland_client.py`, which will help with implementing virtual pointer support.

File: frame/screencopy_tracker.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            shm_pool.destroy()
+
+    def _frame_damage(self, frame, x: int, y: int, width: int, height: int) -> None:
+        self.pending_damage += width * height
+
+    def _frame_ready(self, frame, tv_sec_hi, tv_sec_lo, tv_nsec) -> None:
+        self.frame_count += 1
+        self.total_damage += (self.pending_damage
+            if self.pending_damage
+            else (self.buffer_width * self.buffer_height))
```
Comment (Saviq) [inline frame/screencopy_tracker.py:90]:
```suggestion
        self.total_damage += self.pending_damage or (self.buffer_width * self.buffer_height)
```

File: frame/screencopy_tracker.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        if is_initial:
+            frame.copy(self.buffer)
+        else:
+            frame.copy_with_damage(self.buffer)
+
+    def properties(self) -> dict[str, Any]:
+        total_possible_pixels = max(
+            self.frame_count * self.buffer_width * self.buffer_height,
+            self.total_damage,
+            1
```
Comment (Saviq) [inline frame/screencopy_tracker.py:117]: Why 1?

**Review by Saviq — CHANGES_REQUESTED** [review]
Some nitpicks inline

File: frame/screencopy_tracker.py (hunk 2 above)
Comment (wmww) [reply frame/screencopy_tracker.py:90]: Perhaps too fancy. I think my way is too fancy too. I'm proposing we do it the slightly more verbose and much more obvious way:
```python
if self.pending_damage:
    self.total_damage += self.pending_damage
else:
    self.total_damage += self.buffer_width * self.buffer_height
```
No need to make someone scratch their head of two minutes to save three lines of code.

File: frame/screencopy_tracker.py (hunk 3 above)
Comment (wmww) [reply frame/screencopy_tracker.py:117]: So we don't end up dividing by zero and crashing if nothing gets set

File: frame/screencopy_tracker.py (hunk 2 above)
Comment (Saviq) [reply frame/screencopy_tracker.py]: Your call, the one-liner looks perfectly fine (aka _pythonic_) for me :)

File: frame/screencopy_tracker.py (hunk 3 above)
Comment (Saviq) [reply frame/screencopy_tracker.py]: D'uh, should've looked further down ;)

File: frame/protocols/__init__.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from pathlib import Path
+import pywayland
+import pywayland.scanner
+
+def generate_protocol(name, imports: dict[str, str]) -> dict[str, str]:
+    input_path = Path(__file__).parent.parent / 'data' / f'{name}.xml'
+    proto = pywayland.scanner.Protocol.parse_file(str(input_path))
+    proto_imports = {iface.name: proto.name for iface in proto.interface}
+    output_dir = Path(__file__).parent.parent / 'protocols'
```
Comment (Saviq) [inline frame/protocols/__init__.py:9]: Edited my suggestion too late ;)

```suggestion
    input_path = Path(__file__).parents[1] / 'data' / f'{name}.xml'
    proto = pywayland.scanner.Protocol.parse_file(str(input_path))
    proto_imports = {iface.name: proto.name for iface in proto.interface}
    output_dir = Path(__file__).parents[1] / 'protocols'
```

**Review by Saviq — APPROVED** [review]
Two tinies. Works great!

bors delegate

**Comment by Saviq** [comment]
bors delegate+

Dum dum.

**Pull request merged**